sha2/build.rs

```diff
@@ -3,27 +3,37 @@ fn main() {
 
     let target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap_or_default();
     let target_vendor = env::var("CARGO_CFG_TARGET_VENDOR").unwrap_or_default();
-
+    let target_os = env::var("CARGO_CFG_TARGET_OS").unwrap_or_default();
     let mut build256 = cc::Build::new();
     let (sha256_path, sha512_path) = if target_arch == "x86" {
-        ("src/sha256_x86.S", "src/sha512_x86.S")
+        (["src/sha256_x86.S"].iter(), ["src/sha512_x86.S"].iter())
     } else if target_arch == "x86_64" {
-        ("src/sha256_x64.S", "src/sha512_x64.S")
+        if target_os == "linux" {
+            (
+                ["src/sha256_x64_avx2.S", "src/sha256_x64.S"].iter(),
+                ["src/sha512_x64_avx2.S", "src/sha512_x64.S"].iter(),
+            )
+        } else {
+            (["src/sha256_x64.S"].iter(), ["src/sha512_x64.S"].iter())
+        }
     } else if target_arch == "aarch64" && target_vendor == "apple" {
         build256.flag("-march=armv8-a+crypto");
-        ("src/sha256_aarch64_apple.S", "")
+        (["src/sha256_aarch64_apple.S"].iter(), [""].iter())
     } else if target_arch == "aarch64" {
         build256.flag("-march=armv8-a+crypto");
-        ("src/sha256_aarch64.S", "")
+        (["src/sha256_aarch64.S"].iter(), [""].iter())
     } else {
         panic!("Unsupported target architecture");
     };
```

Comment:
Sorry to keep nitpicking this, but I'd suggest doing this:

```suggestion
   let (sha256_path, sha512_path): (&[&str], &[&str]) = if target_arch == "x86" {
        (&["src/sha256_x86.S"], &["src/sha512_x86.S"])
    } else if target_arch == "x86_64" {
        if target_os == "linux" {
            (
                &["src/sha256_x64_avx2.S", "src/sha256_x64.S"],
                &["src/sha512_x64_avx2.S", "src/sha512_x64.S"],
            )
        } else {
            (&["src/sha256_x64.S"], &["src/sha512_x64.S"])
        }
    } else if target_arch == "aarch64" && target_vendor == "apple" {
        build256.flag("-march=armv8-a+crypto");
        (&["src/sha256_aarch64_apple.S"], &[""])
    } else if target_arch == "aarch64" {
        build256.flag("-march=armv8-a+crypto");
        (&["src/sha256_aarch64.S"], &[""])
    } else {
        panic!("Unsupported target architecture");
    };
```

Reply:
Done